File: src/fenceai/web/static/js/choices.js
// The open questions this plan carries, and the answers it was not built with.
//
// A choice set is the FIFTH kind (spec §3): a hard constraint says *must*, a
// preference says *nicer if*, an objective says *minimise this*, an override
// says *the engine got this wrong here* — and a choice set says **two right
// answers**. Nothing was wrong, so nothing here writes an override; neither
// point is nicer, so nothing here writes a preference. The only honest resolver
// is a person, or the stated default.
//
// **Answering does NOT generate.** Generation stays behind the explicit button
// (spec §16), so a click writes the selection, reloads the PROJECT and marks the
// answer PENDING — the same treatment a dropped post gets. `state.result` is
// left exactly as it was: the plan on screen is still the plan that was built,
// and saying otherwise by silently re-running would be a different lie from the
// one where the row springs back.
//
// **Obligation 5 lives in `valueLabel`.** "Convert units once, at the boundary,
// and keep the source lexeme for display": a footing row published as `24"` is
// shown as `24" (610 mm)`, because a panel showing only our millimetres has
// thrown away the thing a reader checks against — and inventing a lexeme where
// the publisher gave none is worse than showing none.
//
// The two label functions are exported for the node test (`deltaLabel`,
// `valueLabel`) because they are the parts that are wrong at 3 a.m.: a sign, a
// plural, and an obligation. The renderer is DOM and belongs in the smoke suite.
//
// Design: docs/superpowers/specs/2026-09-03-design-choices-and-placement-design.md

import { apiSend, esc } from "./api.js";
import { pushSnapshot } from "./history.js";
import { t } from "./i18n.js";
import { on, reloadProject, state } from "./state.js";
import { sentence, toDisplayValue, tu } from "./units.js";

// Axes are OPEN (spec §5.3) — a point carries only what it differs on — so this
// list is a display ORDER and never a filter. It exists because the backend
// builds `delta` from a Python set intersection, whose iteration order is not
// stable across processes: without a sort the same question would print its
// axes in a different order after a server restart.
const AXIS_ORDER = ["posts", "bays", "boards", "cuts", "pieces",
                     "concrete_l", "holes", "warnings"];

const axisRank = (k) => {
  const i = AXIS_ORDER.indexOf(k);
  return i < 0 ? AXIS_ORDER.length : i;
};

/** What choosing this point would change, relative to what was built.
 *
 *  An EMPTY delta says "no material change" and not nothing: a blank row reads
 *  as a panel that failed to load, and the whole point of the row is that the
 *  two answers cost the same. The minus is U+2212, not a hyphen — these sit
 *  beside numbers in a Hebrew-first UI, and a hyphen-minus there is a dash. */
export function deltaLabel(point) {
  const delta = (point && point.delta) || {};
  const keys = Object.keys(delta)
    .filter((k) => Number(delta[k]) !== 0)
    .sort((a, b) => axisRank(a) - axisRank(b) || (a < b ? -1 : a > b ? 1 : 0));
  if (!keys.length) return t("choices.delta.none");
  return keys.map((k) => axisTerm(k, Number(delta[k]))).join(" · ");
}

function axisTerm(axis, n) {
  const abs = Math.abs(n);
  const key = `choices.delta.${axis}.${abs === 1 ? "one" : "other"}`;
  let word = t(key, { n: abs });
  // An axis the bundles have no word for is still a real measurement — the
  // registry is open. Print the count and the axis name rather than dropping a
  // difference the reader is being asked to weigh.
  if (word === key) word = t("choices.delta.unknown", { n: abs, axis });
  return (n < 0 ? "−" : "+") + word;
}

/** One bound parameter value, in the publisher's words and then in ours.
 *
 *  Contract obligation 5. `24" (610 mm)` where a lexeme exists; a bare
 *  `610 mm` where it does not — never an invented lexeme, which would attribute
 *  a form of words to a publisher who never used it. Our half converts to the
 *  reader's display unit like every other length; the lexeme never does,
 *  because it is a quotation. */
export function valueLabel(point, key) {
  const value = point?.bindings?.[key];
  const lexeme = point?.lexemes?.[key];
  if (!Number.isFinite(Number(value))) return lexeme ? String(lexeme) : "";
  return lexeme
    ? tu("choices.value_with_lexeme", { lexeme, value_mm: value })
    : tu("choices.value", { value_mm: value });
}

// A width list in the reader's display unit. The backend's `label` is raw
// millimetres joined with a dot — correct as an id, wrong as a rendered length
// the moment somebody works in centimetres.
function widthsLabel(point) {
  const widths = point.widths || [];
  if (!widths.length) return String(point.label || "");
  return tu("choices.widths",
            { widths: widths.map((w) => toDisplayValue(w)).join(" · ") });
}

function paramLabel(param) {
  const key = `choices.binding.${param}`;
  const word = t(key);
  return word === key ? param : word;   // unknown parameter: show it raw, never blank
}

// Does this stored answer name this point? A selection carries WHAT was chosen
// (spec §12) — the widths or the bindings — never the name of the generator that
// proposed it, so the comparison is on values.
function pointIsSelected(point, selection) {
  if (!selection) return false;
  const widths = point.widths || [];
  if (widths.length || (selection.widths || []).length)
    return JSON.stringify(widths) === JSON.stringify(selection.widths || []);
  const keys = Object.keys(point.bindings || {});
  if (!keys.length) return false;
  return keys.every((k) => (selection.bindings || {})[k] === point.bindings[k]);
}

function selectionFor(set, selections) {
  return (selections || []).find(
    (c) => c.choice_set === set.id && c.scope === set.scope) || null;
}

/** Render the questions into `host` — and into nothing else.
 *
 *  `sets === null` means there is no run to carry questions, which is a
 *  different sentence from a run that left none. */
export function renderChoices(host, sets, selections) {
  if (!host) return;
  const head = `<h3>${esc(t("choices.title"))}</h3>`;
  if (sets === null || sets === undefined) {
    host.innerHTML = `${head}<div class="meta">${esc(t("choices.no_run"))}</div>`;
    return;
  }
  if (!sets.length) {
    host.innerHTML = `${head}<div class="meta">${esc(t("choices.empty"))}</div>`;
    return;
  }
  host.innerHTML = head
    + `<div class="meta">${esc(t("choices.hint"))}</div>`
    + sets.map((set) => setHtml(set, selectionFor(set, selections))).join("");
}

function setHtml(set, selection) {
  const points = set.points || [];
  const withdraw = selection
    ? `<button type="button" class="choice-withdraw"
         data-set="${esc(set.id)}" data-scope="${esc(set.scope)}"
         >${esc(t("choices.withdraw"))}</button>`
    : "";
  return `<div class="choice-set" data-set="${esc(set.id)}"
      data-scope="${esc(set.scope)}" style="margin-block-start:8px">
    <h4 style="margin:6px 0 2px;font-size:13px">${esc(t(set.question))}</h4>
    <div class="meta" style="color:#64748b;font-size:12px">
      ${sentence("choices.scope", { scope: set.scope })}
      · ${esc(t("choices.options_n", { n: points.length }))}
    </div>
    ${points.map((p) => pointHtml(set, p, selection)).join("")}
    ${withdraw}
  </div>`;
}

function pointHtml(set, point, selection) {
  const chosen = pointIsSelected(point, selection);
  // PENDING is the honest state and the class carries it exactly once: the
  // person answered, the plan on screen was built from the other answer, and
  // nothing will change until they press generate.
  const pending = chosen && !point.is_default;
  const cls = ["choice-point", chosen ? "chosen" : "", pending ? "choice-pending" : ""]
    .filter(Boolean).join(" ");
  const badge = point.is_default
    ? `<span class="tag active">${esc(t("choices.built"))}</span>`
    : chosen ? `<span class="tag proposed">${esc(t("choices.answer_pending"))}</span>`
    : "";
  const by = chosen && selection?.author
    ? `<span class="meta" style="color:#64748b;font-size:11px">${
        esc(t("choices.chosen_by", { author: selection.author }))}</span>`
    : "";
  return `<button type="button" class="${cls}" data-set="${esc(set.id)}"
      data-scope="${esc(set.scope)}" data-point="${esc(point.id)}"
      aria-pressed="${chosen ? "true" : "false"}"
      style="display:block;inline-size:100%;text-align:start;margin:3px 0;
             padding:5px 8px;border:1px solid ${chosen ? "#2563eb" : "#d4dae2"};
             border-radius:6px;background:${chosen ? "#dbeafe" : "#fff"};
             font-size:13px;cursor:pointer">
    <span class="choice-answer">${answerHtml(point)}</span>
    <span class="choice-delta meta"
      style="color:#64748b;font-size:12px;margin-inline-start:6px"
      >${esc(deltaLabel(point))}</span>
    ${badge}${by}
  </button>`;
}

// The figures are LTR-isolated (`.num`) and the words are not: a Hebrew
// parameter label inside a `.num` span reorders on screen.
function answerHtml(point) {
  if ((point.widths || []).length)
    return `<bdi class="num">${esc(widthsLabel(point))}</bdi>`;
  const keys = Object.keys(point.bindings || {});
  if (!keys.length) return `<bdi class="num">${esc(String(point.label || ""))}</bdi>`;
  return keys.map((k) =>
    `${esc(paramLabel(k))} <bdi class="num">${esc(valueLabel(point, k))}</bdi>`
  ).join(" · ");
}

// ---------------------------------------------------------------- the panel

// The panel owns its own host: index.html is not this task's file, and a module
// that reaches into another module's subtree is the thing the module map
// forbids. Created once, beside the site conditions, which is the other
// project-level INPUT to a generation.
function ensureHost() {
  if (typeof document === "undefined") return null;
  let host = document.getElementById("choices");
  if (host) return host;
  const side = document.querySelector(".side-col");
  if (!side) return null;
  host = document.createElement("div");
  host.className = "panel";
  host.id = "choices";
  const after = document.getElementById("site-conditions");
  side.insertBefore(host, after ? after.nextSibling : side.firstChild);
  return host;
}

export function render() {
  const host = ensureHost();
  if (!host) return;
  renderChoices(host,
                state.result ? (state.result.choice_sets || []) : null,
                state.project ? (state.project.choices || []) : []);
  wire(host);
}

function wire(host) {
  for (const el of host.querySelectorAll(".choice-point"))
    el.addEventListener("click",
      () => choose(el.dataset.set, el.dataset.scope, el.dataset.point));
  for (const el of host.querySelectorAll(".choice-withdraw"))
    el.addEventListener("click", () => withdraw(el.dataset.set, el.dataset.scope));
}

function findPoint(setId, scope, pointId) {
  const set = (state.result?.choice_sets || [])
    .find((s) => s.id === setId && s.scope === scope);
  return [set, (set?.points || []).find((p) => p.id === pointId) || null];
}

async function choose(setId, scope, pointId) {
  const [, point] = findPoint(setId, scope, pointId);
  if (!point || !state.projectId) return;
  // pushSnapshot -> mutate -> save, in that order. `project.choices` is IN the
  // snapshot (history.js), so undoing this pops this gesture rather than the
  // drawing edit before it.
  pushSnapshot("choice");
  await apiSend("PUT", `/api/projects/${state.projectId}/choices`, {
    choice_set: setId, scope,
    widths: point.widths || [], bindings: point.bindings || {},
    author: "user",
  });
  // reloadProject, NEVER openProject: this is a non-topology mutation and
  // openProject would wipe the user's undo stack. It also leaves `state.result`
  // alone, which is the whole no-auto-generation rule in one line.
  await reloadProject();
}

async function withdraw(setId, scope) {
  if (!state.projectId) return;
  pushSnapshot("choice-withdraw");
  await apiSend("DELETE",
    `/api/projects/${state.projectId}/choices/${encodeURIComponent(setId)}`
    + `?scope=${encodeURIComponent(scope)}`);
  await reloadProject();
}

export function initChoices() {
  on("project-loaded", render);
  on("result-changed", render);
  on("locale-changed", render);
  on("units-changed", render);   // the same widths, in the reader's unit
  render();
}

File: src/fenceai/strategy/choices.py
"""A choice set: a question the data leaves open, and the points that answer it.

**A fifth kind, deliberately not folded into the other four.** A hard constraint
says *must*; a preference says *nicer if*; an objective says *minimise this*; an
override says *the engine got this wrong here*. A choice set says **two right
answers** — nothing was wrong and neither point is nicer, so the only honest
resolver is a person, or the stated default.

That is also why a selection is anchored to a SCOPE while an override is anchored
to a STATION: an override dies when the fence is redrawn, and a choice should
not. And why a selection is not a CORRECTION (contract obligation 7): a
correction says the engine got it wrong, a selection picks between answers that
are all right.

Pure leaf — `core.units` only. Measuring a point takes a whole generation and
happens in the generator, so this module holds the types and the two rules that
decide what a person is shown, and both fit on one screen.

**Axes are OPEN, and that is the correction of a real mistake.** The first draft
used four fixed measures (posts, boards, cuts, "has an odd bay") and would have
printed *"same posts, same boards, same cuts"* for two footing schedules 25%
apart in concrete — because on a 3 m run both schedules give identical posts and
boards. A point carries the axes it actually differs on.

Design: `docs/superpowers/specs/2026-09-03-design-choices-and-placement-design.md`
"""

from __future__ import annotations

from pydantic import BaseModel

from fenceai.core.units import Mm


class DesignPoint(BaseModel):
    """One admissible answer to a choice set.

    `widths` is a bay list where the point IS a layout; `bindings` are parameter
    values where it asserts some — a `paired` row's depth and span. The two are
    handled differently downstream and the distinction is load-bearing: a
    binding becomes a synthesized `KnowledgeVersion` so `resolve_param` still
    decides, while a width list binds no parameter and needs no synthesizing.
    Reading that the other way round is how a run came to report *"no rule
    states max_span_mm"* about a sealed engineering table.

    `lexemes` carries the source's own words per binding — `24"` beside 610 —
    because contract obligation 5 is *"convert units once, at the boundary, and
    keep the source lexeme for display."* A panel showing only our millimetres
    has thrown away what a reader checks against.

    `axes` are physical counts measured BY a probe: never money. A stored price
    goes stale the moment the catalog moves, and ADR-0011 puts what a fence
    costs in a `SupplyRun` against one yard.

    `is_default` marks what the engine builds. A default is never eliminated.
    """

    id: str
    label: str
    widths: list[Mm] = []
    bindings: dict[str, Mm] = {}
    lexemes: dict[str, str] = {}
    axes: dict[str, int] = {}
    # What choosing this would change relative to the baseline, on shared axes.
    delta: dict[str, int] = {}
    is_default: bool = False


class ChoiceSet(BaseModel):
    """A question, its points, and what it depends on.

    `scope` is a GAP between fixed stations, not a section. Any corner, gate
    edge, terrain step or single pinned post makes a run several gaps, and a
    section-scoped answer would be applied to a gap it was never measured for —
    which is a wrong fence, and self-defeating besides: the moment a person pins
    a post the run becomes two gaps.

    `depends_on` names a set whose answer must settle first. The placement
    question exists only while the chosen widths leave a stub, and a dependent
    set whose parent moved is dropped rather than orphaned.
    """

    id: str
    scope: str
    question: str
    points: list[DesignPoint] = []
    depends_on: str | None = None

    def default(self) -> DesignPoint | None:
        """The point the engine built, which is always present."""
        return next((p for p in self.points if p.is_default), None)


def dominates(a: DesignPoint, b: DesignPoint) -> bool:
    """Is `a` at least as good as `b` on every axis BOTH carry, and better on one?

    **Only shared axes are compared.** Treating a missing axis as zero would let
    a footing point measured in concrete dominate a layout point measured in
    boards, on an axis the second never claimed.

    The strict improvement is what stops two points measuring the same from each
    eliminating the other and leaving nothing offered.

    Every axis minimises, which is exactly why "has an odd bay" is not an axis:
    it is taste, and taste does not eliminate. It is printed on the row instead.
    Keeping it as a filter axis was the only thing hiding that this engine's own
    default layout is dominated on posts, boards and cuts.
    """
    shared = set(a.axes) & set(b.axes)
    if not shared:
        return False
    return (all(a.axes[k] <= b.axes[k] for k in shared)
            and any(a.axes[k] < b.axes[k] for k in shared))


def offered(points: list[DesignPoint]) -> list[DesignPoint]:
    """The points worth showing, in the order they were generated.

    **The default is exempt.** Without that the layout the engine builds can be
    absent from its own panel — and a stored selection naming it would then
    report `choice_unavailable` for a layout being built on the same screen.
    That one rule also retires a `min_span` rule which only warns in
    `layout_segment` while rejecting in candidate generation, and a
    `prefer_equal=False` baseline that no generator proposes.

    Order is preserved rather than sorted by any axis: sorting would make the
    panel's first row a judgement this module has no basis for, and the
    generators' order is already meaningful — the default, then the manufactured
    tiling, then yield.
    """
    return [
        p for p in points
        if p.is_default
        or not any(o is not p and dominates(o, p) for o in points)
    ]
